### PR TITLE
consensus: add dBFT watchdog

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Requested by @realloc.

#### Problem

dBFT is not perfect, and sometimes it hangs in a deadlock.

#### Solution

Add dBFT watchdog. Watchdog tracks time spent since the last accepted block and enforses the consensus reset on threshold.

No documentation update included. If we'll decide to include this feature into master, then a separate PR will be created.

Depends on #2567 and #2566.

### CODE REVIEW

File: pkg/consensus/consensus.go [hunk 1]
```diff
@@ -292,6 +303,27 @@ events:
 			s.log.Debug("timer fired",
 				zap.Uint32("height", hv.Height),
 				zap.Uint("view", uint(hv.View)))
+			if s.Config.ProtocolConfiguration.EnableDBFTWatchdog {
```
Comment (fyrchik) [inline pkg/consensus/consensus.go:306]: I think we need to use another timer which starts with new block approval. It should depend on `SecondsPerBlock`, last block timestamp and the current time.

File: pkg/consensus/consensus.go [hunk 2]
```diff
@@ -292,6 +303,27 @@ events:
...
+				cvWatcher.mtx.Lock()
```
Comment (fyrchik) [inline pkg/consensus/consensus.go:307]: Why do we need a mutex here?

File: pkg/consensus/consensus.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/consensus/consensus.go]: Fixed. But this way of consensus restart is also imperfect, because timer can't guarantee simultaneous node restart:
```
Timer that will send the current time on its channel after at least duration d.
```

**Review by roman-khimov — CHANGES_REQUESTED** [review]
NACK. This goes against dBFT core logic and is very likely to produce forks or some other unpredictable problems. We have #2552 to properly investigate and fix dBFT problems if any.

**Comment by realloc** [comment]
The meaning of the watchdog is to prevent any chain halt situation that we couldn't foresee. That's the requirement for on-prem installs and it's not related to #2552 at all.

In on-prem installs, where neo-go is used as a database, we can't restart neo-go instance by external  system as it affects the whole stack operations. At least 3 times on prod we had a situation with locked consensus mechanism causing the whole cluster malfunction, which was fixed by simultaneous neo-go nodes restart. To prevent such situation we need this kind of watchdog doing a simultaneous consensus mechanism restart without affecting the RPC operations. As we discussed with @fyrchik and @AnnaShaleva, the restart could be tied to the nearest round number minute or similar point in time.

As we need this anyway, let's try to also attach some instructions for operations people on how to gather the most of useful debug information when they hit the consensus halt situation again.

File: pkg/consensus/consensus.go [hunk 3]
```diff
@@ -292,6 +303,27 @@ events:
...
+				if hv.Height != cvWatcher.height {
+					cvWatcher.height = hv.Height
+					cvWatcher.count = 0
+				}
+				cvWatcher.count++
+				if cvWatcher.count > s.Config.ProtocolConfiguration.DBFTWatchdogThreshold*numOfCNs {
+					var newView byte = 0
+					s.log.Warn("couldn't accept new block, restarting consensus service",
```
[the remark this answers is not in the document]
Comment (realloc) [reply pkg/consensus/consensus.go]: Can we have a counter with a height label? I hope will not eat much memory.

**Comment by roman-khimov** [comment]
> we can't restart neo-go instance by external system
> was fixed by simultaneous neo-go nodes restart

Hmm.

My presumption is that all of the services are monitored anyway and if they are, external entities can do whatever they want if needed.

> we had a situation with locked consensus mechanism

dBFT runs over the internet on chains with many millions of blocks. We know it's not perfect, but it works more often than not. If we have any problem with dBFT it needs to be properly investigated and fixed.

The problem with this PR is that in fact we don't know how it will affect the system at all. Each node has a clock and a timer of its own, it's not the same as when you externally restart all of them within very short time frame.

**Comment by realloc** [comment]
As it is written in the code comment, we need this option to be toggled in on-prem environments only. 

> My presumption is that all of the services are monitored anyway and if they are, external entities can do whatever they want if needed.

For the target environment this presumption is wrong.

>dBFT runs over the internet on chains with many millions of blocks. We know it's not perfect, but it works more often than not. If we have any problem with dBFT it needs to be properly investigated and fixed.

Again, this is irrelevant for the target case.

> The problem with this PR is that in fact we don't know how it will affect the system at all. 

It will give a change for the isolated system to self recover after the unforeseen problem. We don't plan to use it in public CN setup

> Each node has a clock and a timer of its own, it's not the same as when you externally restart all of them within very short time frame.

Local time is NTP controlled and there is no external system to restart the service. Again, we don't want the full restart because other services will read the chain actively and with this watchdog we can avoid unnecessary service interrupt.

File: cli/server/server.go [hunk 4]
```diff
@@ -405,7 +405,7 @@ func mkOracle(config network.ServerConfig, chain *core.Blockchain, serv *network
 	return orc, nil
 }
 
-func mkConsensus(config network.ServerConfig, chain *core.Blockchain, serv *network.Server, log *zap.Logger) (consensus.Service, error) {
+func mkConsensus(config network.ServerConfig, chain *core.Blockchain, serv *network.Server, log *zap.Logger, restartCh chan struct{}) (consensus.Service, error) {
```
Comment (AnnaShaleva) [inline cli/server/server.go:408]: @roman-khimov, not sure whether we need a channel for consensus restart. We may go with callback if needed.

Also, current PR logic allows to move consensus reloading completely outside of consensus server code.

**Comment by AnnaShaleva** [comment]
Ready for review, testing is in progress.

File: pkg/consensus/consensus.go [hunk 5]
```diff
@@ -334,6 +360,33 @@ events:
...
+			break
+		}
+	}
+	if s.watchdogTimer == nil {
+		// Don't reinitialize consensus for the first time, the service has just started.
+		s.watchdogTimer = time.NewTimer(resetAfter)
+	} else {
+		s.watchdogTimer.Reset(resetAfter)
+		if delta+threshold <= 0 {
+			s.log.Warn("couldn't accept new block, sending signal to restart consensus service",
```
Comment (realloc) [inline pkg/consensus/consensus.go:380]: To simplify grep, could we use something like "DBFTWatchdog triggered after X sec from last height Y.  Sending signal to restart consensus service."

File: pkg/consensus/consensus.go [hunk 6]
```diff
@@ -286,6 +286,7 @@ events:
 		select {
 		case <-s.quit:
 			s.dbft.Timer.Stop()
+			s.Chain.UnsubscribeFromBlocks(s.blockEvents)
```
Comment (roman-khimov) [inline pkg/consensus/consensus.go:309]: It'd be nice to have this in a separate PR to be merged into the `master` branch.

File: pkg/consensus/consensus.go [hunk 7]
```diff
@@ -281,12 +285,31 @@ func (s *service) Shutdown() {
...
+	var watchdogThreshold time.Duration
+	if s.Config.ProtocolConfiguration.EnableDBFTWatchdog {
+		if s.lastTimestamp == 0 {
+			curr, err := s.Chain.GetBlock(s.Chain.CurrentBlockHash())
+			if err != nil {
+				s.log.Error("failed to set last block timestamp",
+					zap.Error(err))
+				close(s.finished)
+			}
+			s.lastTimestamp = curr.Timestamp
```
Comment (roman-khimov) [inline pkg/consensus/consensus.go:297]: Block 0 time will be fun wrt this logic. Rare event though.

File: pkg/consensus/consensus.go [hunk 8]
```diff
@@ -335,6 +360,32 @@ events:
...
 }
 
+func (s *service) resetWatchdogTimer(threshold time.Duration) {
+	lastTimestamp := int64(s.lastTimestamp)
+	now := time.Now().UnixMilli()
+	delta := time.Millisecond * time.Duration(lastTimestamp-now)
+	resetAfter := delta
+	for {
+		resetAfter += threshold
+		if resetAfter > 0 {
```
Comment (roman-khimov) [inline pkg/consensus/consensus.go:370]: Huh, this solves block 0 btw.

**Review by roman-khimov — COMMENTED** [review]
Does this work for more than 15m? I think you should `resetWatchdogTimer()` after every block and I don't see any code doing this.

File: cli/server/server.go [hunk 9]
```diff
@@ -538,8 +540,20 @@ Main:
 					rpcServer.Start()
 				}
 			}
+		case <-consensusRestartChan:
+			log.Warn("restart signal is received from consensus service, restarting consensus service")
+			consensusService.Shutdown()
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/server/server.go:548]: Nothing good, I'll think about it.

File: pkg/consensus/consensus.go (hunk 8 above)
Comment (AnnaShaleva) [reply pkg/consensus/consensus.go:370]: Yes, it was done to solve this problem. And to be able to implement [this optimisation](https://github.com/nspcc-dev/neo-go/pull/2562#issuecomment-1168640993) also. I think it's OK to have this overhead for genesis block, it happens once per the whole node lifetime. Or do we need a separate logic for genesis?

**Comment by roman-khimov** [comment]
> reset timer wrt N+k timestamp and current time to reach overall 15-minutes threshold

I think it'd be more clear if this timer only expired when there is a need to reset. Otherwise it would be restarted with every block. At least that's what I'd expect from a watchdog-like thing, even though I can't call this one a watchdog.

File: pkg/consensus/consensus.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/consensus/consensus.go:289]: https://github.com/nspcc-dev/neo-go/pull/2566.

File: pkg/consensus/consensus.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/consensus/consensus.go]: I'm not sure it's a good idea to put `X` and `Y` right into the log message. Can we keep them as log fields?

Currently, logs look like
```
2022-06-28T15:18:42.921Z	WARN	couldn't accept new block, sending signal to restart consensus service	{"latest block index": 18, "latest block timestamp": 1656429502920, "time since latest block": "20.001s", "time until next restart": "9.999s"}
```
Isn't it enough for grepping?

Comment (realloc) [reply pkg/consensus/consensus.go]: Yes, looks good. Missed the lines following the log message =)

**Comment by AnnaShaleva** [comment]
Rebased onto the current master.

File: pkg/consensus/watchdog.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
+package consensus
+
+import (
+	"errors"
+	"time"
+
+	coreb "github.com/nspcc-dev/neo-go/pkg/core/block"
+
```
Comment (roman-khimov) [inline pkg/consensus/watchdog.go:8]: Oopsie.

**Review by roman-khimov — COMMENTED** [review]
Nice kludge, services are cool.

File: pkg/consensus/watchdog.go (hunk 10 above)
Comment (AnnaShaleva) [reply pkg/consensus/watchdog.go]: Fixed.

File: cli/server/server.go [hunk 11]
```diff
@@ -538,9 +558,26 @@ Main:
...
+			if err != nil {
+				shutdownErr = fmt.Errorf("failed to create new instance of consensus service: %w", err)
+				cancel()
+			}
+			// After that, gracefully shutdown the old consensus service and replace it by the new one.
+			consensusService.Shutdown()
+			consensusService = newConsensusService
+			// Start new consensus service irrespective to serv.IsInSync(), because if server is temporary out of sync
+			// then consensus service will never be started by serv. See https://github.com/nspcc-dev/neo-go/issues/2564.
+			consensusService.Start()
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/server/server.go:579]: We have the same situation if node is temporary out of sync. dBFT can handle stale messages, so let's keep this code until https://github.com/nspcc-dev/neo-go/issues/2564 is fixed.

File: cli/server/server.go [hunk 12]
```diff
@@ -538,9 +558,26 @@ Main:
 					rpcServer.Start()
 				}
 			}
+		case <-consensusRestartChan:
+			log.Info("restart signal is received from consensus watchdog service, restarting consensus service")
+			// Firstly, create new consensus so that network server is able to use proper OnPayload and OnTransaction callbacks.
+			// At this stage old consensus service is still able to send consensus messages to other nodes via BroadcastExtensible
+			// and request transactions via RequestTx, but it's OK.
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply cli/server/server.go:565]: 😩

**Comment by AnnaShaleva** [comment]
It works somehow, at least peers disconnected before the consensus reloading and trying to connect again after new consensus instance is created.
```
2022-06-29T12:27:54.737Z	INFO	request change view	{"view": 0, "height": 37, "reason": "Timeout", "new_view": 1, "nc": 0, "nf": 1}
2022-06-29T12:27:54.738Z	INFO	received ChangeView	{"validator": 0, "reason": "Timeout", "new view": 1}
2022-06-29T12:27:54.999Z	INFO	new peer connected	{"addr": "172.200.0.2:51942", "peerCount": 6}
2022-06-29T12:27:55.722Z	WARN	couldn't accept new block, sending signal to restart consensus service	{"latest block index": 36, "latest block timestamp": 1656505660722, "time since latest block": "15s", "time till next restart": "15s"}
2022-06-29T12:27:55.722Z	INFO	restart signal is received from consensus watchdog service, restarting consensus service
2022-06-29T12:27:56.082Z	WARN	peer disconnected	{"addr": "172.200.0.2:51818", "error": "unexpected empty payload: CMDVersion", "peerCount": 5}
2022-06-29T12:27:56.086Z	WARN	peer disconnected	{"addr": "172.200.0.2:20334", "error": "dogs bite", "peerCount": 4}
2022-06-29T12:27:56.086Z	WARN	peer disconnected	{"addr": "172.200.0.2:20334", "error": "dogs bite", "peerCount": 3}
2022-06-29T12:27:56.086Z	WARN	peer disconnected	{"addr": "172.200.0.2:20334", "error": "dogs bite", "peerCount": 2}
2022-06-29T12:27:56.086Z	WARN	peer disconnected	{"addr": "172.200.0.2:51934", "error": "dogs bite", "peerCount": 1}
2022-06-29T12:27:56.086Z	INFO	starting consensus service
2022-06-29T12:27:56.086Z	WARN	peer disconnected	{"addr": "172.200.0.2:51942", "error": "dogs bite", "peerCount": 0}
2022-06-29T12:27:56.086Z	INFO	initializing dbft	{"height": 37, "view": 0, "index": 2, "role": "Backup"}
2022-06-29T12:27:58.086Z	INFO	skip change view	{"nc": 0, "nf": 3}
2022-06-29T12:27:59.087Z	INFO	new peer connected	{"addr": "172.200.0.2:20334", "peerCount": 1}
2022-06-29T12:28:02.088Z	WARN	peer disconnected	{"addr": "172.200.0.2:20334", "error": "unexpected empty payload: CMDVersion", "peerCount": 0}
2022-06-29T12:28:02.088Z	INFO	new peer connected	{"addr": "172.200.0.2:20334", "peerCount": 1}
2022-06-29T12:28:02.088Z	INFO	skip change view	{"nc": 0, "nf": 3}
2022-06-29T12:28:02.088Z	INFO	new peer connected	{"addr": "172.200.0.2:51952", "peerCount": 2}
2022-06-29T12:28:02.088Z	INFO	new peer connected	{"addr": "172.200.0.1:20333", "peerCount": 3}
2022-06-29T12:28:02.089Z	INFO	new peer connected	{"addr": "172.200.0.1:47376", "peerCount": 4}
2022-06-29T12:28:02.089Z	INFO	started protocol	{"addr": "172.200.0.2:20334", "userAgent": "/NEO-GO:0.99.1-pre-69-gb21a3c99/", "startHeight": 36, "id": 2730790704}
2022-06-29T12:28:02.089Z	INFO	new peer connected	{"addr": "172.200.0.2:20334", "peerCount": 5}
2022-06-29T12:28:02.090Z	WARN	peer disconnected	{"addr": "172.200.0.1:20333", "error": "identical node id", "peerCount": 4}
2022-06-29T12:28:02.090Z	INFO	new peer connected	{"addr": "172.200.0.2:51966", "peerCount": 5}
2022-06-29T12:28:02.090Z	INFO	new peer connected	{"addr": "172.200.0.2:51968", "peerCount": 6}
2022-06-29T12:28:02.090Z	INFO	new peer connected	{"addr": "172.200.0.2:20334", "peerCount": 7}
2022-06-29T12:28:02.090Z	INFO	new peer connected	{"addr": "172.200.0.1:20333", "peerCount": 8}
2022-06-29T12:28:02.090Z	INFO	new peer connected	{"addr": "172.200.0.1:47394", "peerCount": 9}
```

File: cli/server/server.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply cli/server/server.go:579]: I don't think it's the same, in normal scenario we never restart consensus, so once it's up (when the node is in sync) we don't care about `IsInSync()`. Here we can end up with starting it before the node ever reached the sync. It's not likely to happen though.

**Comment by roman-khimov** [comment]
1. This branch is outdated and can't be merged immediately.
2. It's been 6 months since this experiment and we had no issues that would require this.
3. This logic breaks the basic dBFT assumption of `f` misbehaving nodes, we suddenly can have _all_ nodes misbehaving and this can lead to unpredictable results.
4. Any other potential dBFT problems are a part of #2552 and will be investigated/fixed there properly.
